src/kernels/MIOpenMultilabelSoftMarginLoss.cpp

```diff
@@ -0,0 +1,130 @@
+                                                           tensor_view_t<2> T_tv,
+                                                           tensor_view_t<1> W_tv,
+                                                           tensor_view_t<1> O_tv)
+{
+    const uint64_t gid = threadIdx.x + blockIdx.x * blockDim.x;
+
+    size_t N = I_tv.size[0], C = I_tv.size[1];
+    size_t n = gid;
+    if(n >= N)
+        return;
+
+    FLOAT_ACCUM loss = 0;
+
+    // TODO: maybe faster reduce sum here
+    for(size_t c = 0; c < C; c++)
+    {
+        FLOAT_ACCUM w = CVT_FLOAT2ACCUM(W[W_tv.get_tensor_view_idx({c})]);
+        FLOAT_ACCUM i = CVT_FLOAT2ACCUM(I[I_tv.get_tensor_view_idx({n, c})]);
+        FLOAT_ACCUM t = CVT_FLOAT2ACCUM(T[T_tv.get_tensor_view_idx({n, c})]);
+
+        loss += -w * calc_loss(i, t);
+    }
+
+    O[O_tv.get_tensor_view_idx({n})] = CVT_ACCUM2FLOAT(loss / C);
+}
+
+extern "C" __global__ void
+MultilabelSoftMarginLossUnreducedForward2d(const INPUT_TYPE* __restrict__ I,
+                                           const INPUT_TYPE* __restrict__ T,
+                                           const INPUT_TYPE* __restrict__ W,
+                                           INPUT_TYPE* __restrict__ O,
+                                           tensor_view_t<2> I_tv,
+                                           tensor_view_t<2> T_tv,
+                                           tensor_view_t<1> W_tv,
+                                           tensor_view_t<1> O_tv)
+{
+    // instantiate the kernel
+    multilabelsoftmarginlossunreducedforward2d<INPUT_TYPE>(I, T, W, O, I_tv, T_tv, W_tv, O_tv);
+}
+
+template <typename DTYPE>
+__device__ void multilabelsoftmarginlossforward2d(const DTYPE* __restrict__ I,
+                                                  const DTYPE* __restrict__ T,
+                                                  const DTYPE* __restrict__ W,
+                                                  DTYPE* __restrict__ lsum,
+                                                  const float divisor,
+                                                  tensor_view_t<2> I_tv,
+                                                  tensor_view_t<2> T_tv,
+                                                  tensor_view_t<1> W_tv)
+{
+    const uint64_t gid = threadIdx.x + blockIdx.x * blockDim.x;
+
+    size_t N = I_tv.size[0], C = I_tv.size[1];
+    size_t n = gid;
+
+    if(n >= N)
+        return;
+
+    FLOAT_ACCUM loss = 0;
+
+    for(size_t c = 0; c < C; c++)
+    {
+        FLOAT_ACCUM w = CVT_FLOAT2ACCUM(W[W_tv.get_tensor_view_idx({c})]);
+        FLOAT_ACCUM i = CVT_FLOAT2ACCUM(I[I_tv.get_tensor_view_idx({n, c})]);
+        FLOAT_ACCUM t = CVT_FLOAT2ACCUM(T[T_tv.get_tensor_view_idx({n, c})]);
+
+        loss += -w * calc_loss(i, t);
+    }
+
+    lsum[n] = CVT_ACCUM2FLOAT(loss / C / divisor);
+}
+
+extern "C" __global__ void MultilabelSoftMarginLossForward2d(const INPUT_TYPE* __restrict__ I,
```

Comment:
Same as above

Reply:
Fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/e9f959f5c1bfb65a1fb4033449671487b45c9bf0)